common4j/src/main/com/microsoft/identity/common/java/net/UrlConnectionHttpClient.java

```
@@ -0,0 +1,361 @@
+            safeCloseStream(responseStream);
+        }
+
+        return response;
+    }
+
+    private HttpURLConnection setupConnection(HttpRequest request) throws IOException {
+        final HttpURLConnection urlConnection = HttpUrlConnectionFactory.createHttpURLConnection(request.getRequestUrl());
+
+        // Apply request headers and update the headers with default attributes first
+        final Set<Map.Entry<String, String>> headerEntries = request.getRequestHeaders().entrySet();
+
+        for (final Map.Entry<String, String> entry : headerEntries) {
+            urlConnection.setRequestProperty(entry.getKey(), entry.getValue());
+        }
+
+        urlConnection.setConnectTimeout(getConnectTimeoutMs());
+        urlConnection.setReadTimeout(getReadTimeoutMs());
+        urlConnection.setInstanceFollowRedirects(true);
+        urlConnection.setUseCaches(false);
+        urlConnection.setDoInput(true);
+
+        return urlConnection;
+    }
+
+    private Integer getReadTimeoutMs() {
+        return readTimeoutMsSupplier == null ? readTimeoutMs : readTimeoutMsSupplier.get();
+    }
+
+    private Integer getConnectTimeoutMs() {
+        return connectTimeoutMsSupplier == null ? connectTimeoutMs : connectTimeoutMsSupplier.get();
+    }
+
+    private static void setRequestBody(@NonNull final HttpURLConnection connection,
+                                       final byte[] contentRequest,
+                                       final String requestContentType) throws IOException {
+        if (contentRequest == null) {
+            return;
+        }
+
+        connection.setDoOutput(true);
+
+        if (!StringUtil.isNullOrEmpty(requestContentType)) {
+            connection.setRequestProperty("Content-Type", requestContentType);
+        }
+
+        connection.setRequestProperty("Content-Length", String.valueOf(contentRequest.length));
+
+        OutputStream out = null;
+
+        try {
+            out = connection.getOutputStream();
```

>out = connection.getOutputStream(); [](http://example.com/codeflow?start=12&length=35)

if this fails or throws.. does it set the response/input stream accordingly?
Do we have test cases covering that?

---

was just looking at this.. [https://stackoverflow.com/questions/4816824/httpurlconnection-post-conn-getoutputstream-throwing-exception](https://stackoverflow.com/questions/4816824/httpurlconnection-post-conn-getoutputstream-throwing-exception)

---

We're currently throwing IOException, which is base class of the SocketException.

Those exceptions are not handled in this layer.